### ISSUE TITLE
feat: add top-p probability distribution database infrastructure

### ISSUE THREAD

**Issue by jacob-danner** [jacob-danner/dissecting-vlm#1]
**Work Until Complete**:
   - This is an autonomous task - keep working until PR is created OR you encounter an unrecoverable error
   - Debug and fix issues iteratively
   - If validation fails, fix the issues and re-run
   - Only stop when: (a) PR successfully created, OR (b) fundamental blocker that requires human intervention
**Required Workflow**:
   1. Verify on feat/top-p-database branch (`git branch --show-current`)
   2. Implement all files (database.py, populate_database.py, validate_database.py, edit vlm_tools.py)
   3. Test with `uv run python populate_database.py`
   4. Validate with `uv run python validate_database.py`
   5. If validation fails → debug, fix, re-run (iterate until it passes)
   6. Only create PR if validation passes (exit code 0)
   7. PR title: "feat: add top-p probability distribution database infrastructure"
   8. Include validation output in PR description

## Objective

Extend the existing VLM unembedding analysis to capture **full probability distributions** (top-p) instead of just argmax. Build a DuckDB database to store these distributions for systematic positional analysis.

## Background

Current implementation (`vlm_tools.py::dissect_image`) performs greedy nearest-neighbor decoding (argmax) to map each of 256 image tokens to a single vocabulary token. This is lossy - the full probability distribution contains much more semantic information.

**Goal**: Capture top-p probability distributions for each of the 256 image token positions and store them in a queryable database for downstream analysis.

### 1. Database Schema (DuckDB)

Create a normalized schema with two tables:

#### Table 1: `image_token_distributions`
Stores the top-p probability distribution for each image token position.

```sql
CREATE TABLE image_token_distributions (
    image_id VARCHAR NOT NULL,           -- Identifier for the image (e.g., "monocolor.png")
    token_position INTEGER NOT NULL,     -- Position in sequence (0-255)
    token_id INTEGER NOT NULL,           -- Vocabulary token ID
    token_text VARCHAR NOT NULL,         -- Decoded token text (e.g., " the", "mountain")
    probability FLOAT NOT NULL,          -- Probability of this token at this position
    probability_rank INTEGER NOT NULL,   -- Rank when sorted by probability DESC for this position
    PRIMARY KEY (image_id, token_position, token_id)
);
```

**Notes**:
- Each row represents one token in the top-p distribution for a given (image, position)
- A single token_position will have multiple rows (as many tokens as needed to reach top-p threshold)
- `probability_rank`: 1 = highest probability (argmax), 2 = second highest, etc.

#### Table 2: `image_token_embeddings`
Stores the actual embedding vectors being unembedded (for reference/future use).

```sql
CREATE TABLE image_token_embeddings (
    image_id VARCHAR NOT NULL,
    token_position INTEGER NOT NULL,
    embedding_vector FLOAT[2560] NOT NULL,  -- d_model for Gemma 3
    PRIMARY KEY (image_id, token_position)
);
```

**Notes**:
- Stores the 2560-dimensional vector (d_model) for each image token position
- Can be used for future analyses (e.g., computing cosine similarities, visualizations)
- Use DuckDB's native array type - convert torch tensors with `.cpu().numpy().tolist()`

### 2. Code Implementation

**File structure**:
- Extend or create new functions in `vlm_tools.py` (match existing code style)
- Create `populate_database.py` as executable script (match repo conventions)
- Create `database.py` for database operations
- Create `validate_database.py` for validation checks

**Implementation notes**:
- Use MPS device when available for inference speed
- Image paths: Relative to repo root (e.g., `images/monocolor.png`)
- Image IDs: Use filename only (e.g., `monocolor.png`) for database portability
- Database handling: Overwrite existing database on each run (drop/recreate tables)
- Top-p tokens: Include all tokens needed to reach threshold, no artificial limits

##### In `vlm_tools.py`:

```python
def extract_image_token_distributions(
    image_path: str,
    model: Gemma3ForConditionalGeneration,
    processor: AutoProcessor,
    top_p: float = 0.9
) -> tuple[torch.Tensor, list[dict]]:
    """
    Extract image tokens and compute top-p probability distributions.

    Args:
        image_path: Path to image file
        model: Gemma3 model
        processor: Gemma3 processor
        top_p: Cumulative probability threshold (default 0.9)

    Returns:
        embeddings: Tensor of shape (256, d_model) - the image token embeddings
        distributions: List of 256 dicts, each containing:
            {
                'position': int,
                'tokens': list of (token_id, token_text, probability, rank) tuples
            }
    """
    # 1. Extract image tokens (reuse existing extract_image_tokens logic, adjust it as needed while not breaking how existing consumers use it)
    # 2. For each token, pass through lm_head to get logits
    # 3. Apply softmax to get probabilities
    # 4. For each position, extract top-p tokens (cumulative probability >= top_p)
    # 5. Rank tokens by probability
    pass
```

**Key implementation details**:
- Reuse existing `extract_image_tokens` logic for getting the 256 image embeddings
- For each embedding: `logits = model.lm_head(embedding)`
- Get probabilities: `probs = torch.softmax(logits, dim=-1)`
- Implement top-p filtering:
  ```python
  sorted_probs, sorted_indices = torch.sort(probs, descending=True)
  cumulative_probs = torch.cumsum(sorted_probs, dim=0)
  top_p_mask = cumulative_probs <= top_p
  # Include one more token to ensure we exceed top_p threshold
  top_p_mask[top_p_mask.sum()] = True
  ```
- Decode token_ids to text: `processor.tokenizer.decode(token_id.item())`

##### In `database.py`:

```python
import duckdb
from pathlib import Path

def create_database(db_path: str = "vlm_analysis.db") -> None:
    """
    Create DuckDB database with required schema.

    Args:
        db_path: Path to database file
    """
    # Create tables with schema defined above
    pass

def insert_image_distributions(
    db_path: str,
    image_id: str,
    distributions: list[dict],
    embeddings: torch.Tensor
) -> None:
    """
    Insert image token distributions and embeddings into database.

    Args:
        db_path: Path to database file
        image_id: Unique identifier for image
        distributions: Output from extract_image_token_distributions
        embeddings: Image token embedding vectors (256, d_model)
    """
    # Insert into image_token_distributions
    # Insert into image_token_embeddings
    pass

def query_position_tokens(
    db_path: str,
    position: int,
    top_k: int = 10
) -> list[dict]:
    """
    Query most common tokens at a given position across all images.

    Args:
        db_path: Path to database file
        position: Token position (0-255)
        top_k: Number of top tokens to return

    Returns:
        List of {token_text, avg_probability, frequency} dicts
    """
    # Example analysis query for validation
    pass
```

##### In `populate_database.py`:

```python
"""
Populate the VLM analysis database with all repo images.
"""

from vlm_tools import load_model_and_processor, extract_image_token_distributions
from database import create_database, insert_image_distributions
from pathlib import Path

def main():
    # Image paths
    images = [
        "images/monocolor.png",
        "images/bicolor.png",
        "images/sunrise_mountain.jpg",
        "images/chicago_l.jpg",
        "images/code_snippet.png",
        "images/vision_language_model_data_flow.png"
    ]

    # Initialize
    db_path = "vlm_analysis.db"
    create_database(db_path)
    model, processor = load_model_and_processor()

    # Process each image
    for image_path in images:
        print(f"Processing {image_path}...")
        image_id = Path(image_path).name
        embeddings, distributions = extract_image_token_distributions(
            image_path, model, processor, top_p=0.9
        )
        insert_image_distributions(db_path, image_id, distributions, embeddings)
        print(f"  Inserted {sum(len(d['tokens']) for d in distributions)} total tokens")

    print(f"\nDatabase populated: {db_path}")

if __name__ == "__main__":
    main()
```

##### In `validate_database.py`:

```python
#!/usr/bin/env python3
"""
Validate the VLM analysis database to ensure data integrity.
Runs all validation checks and reports pass/fail.
"""

import duckdb
import sys

def validate_database(db_path: str = "vlm_analysis.db") -> bool:
    """
    Run all validation checks on the database.

    Args:
        db_path: Path to database file

    Returns:
        True if all validations pass, False otherwise
    """
    conn = duckdb.connect(db_path, read_only=True)
    all_passed = True

    print("Running validation checks...\n")

    # Check 1: Row count check
    print("1. Row count check:")
    result = conn.execute("""
        SELECT image_id, COUNT(*) as num_tokens
        FROM image_token_distributions
        GROUP BY image_id
        ORDER BY image_id;
    """).fetchall()

    for image_id, count in result:
        print(f"   {image_id}: {count} tokens")
        if count < 256 or count > 50000:  # Sanity bounds
            print(f"   ⚠️  WARNING: Unexpected token count")
            all_passed = False
    print("   ✓ Row counts within expected range\n")

    # Check 2: Probability sum check
    print("2. Probability sum check:")
    invalid = conn.execute("""
        SELECT image_id, token_position, SUM(probability) as prob_sum
        FROM image_token_distributions
        GROUP BY image_id, token_position
        HAVING prob_sum < 0.89 OR prob_sum > 1.0;
    """).fetchall()

    if invalid:
        print(f"   ✗ FAILED: {len(invalid)} positions have invalid probability sums")
        for row in invalid[:5]:  # Show first 5
            print(f"     {row}")
        all_passed = False
    else:
        print("   ✓ All positions have valid probability sums (≥0.9, ≤1.0)\n")

    # Check 3: Ranking consistency check
    print("3. Ranking consistency check:")
    # Check that probability_rank matches actual probability ordering
    invalid_ranks = conn.execute("""
        WITH ranked AS (
            SELECT
                image_id,
                token_position,
                probability,
                probability_rank,
                ROW_NUMBER() OVER (
                    PARTITION BY image_id, token_position
                    ORDER BY probability DESC
                ) as expected_rank
            FROM image_token_distributions
        )
        SELECT COUNT(*) as mismatches
        FROM ranked
        WHERE probability_rank != expected_rank;
    """).fetchone()[0]

    if invalid_ranks > 0:
        print(f"   ✗ FAILED: {invalid_ranks} rows have incorrect probability_rank")
        all_passed = False
    else:
        print("   ✓ All probability_rank values match probability ordering\n")

    # Check 4: Embedding vectors check
    print("4. Embedding vectors check:")
    embedding_count = conn.execute("""
        SELECT COUNT(*) FROM image_token_embeddings;
    """).fetchone()[0]

    expected_embeddings = 6 * 256  # 6 images × 256 positions
    if embedding_count != expected_embeddings:
        print(f"   ✗ FAILED: Expected {expected_embeddings} embeddings, found {embedding_count}")
        all_passed = False
    else:
        print(f"   ✓ All {embedding_count} embeddings stored\n")

    # Check 5: Data completeness - all images × positions present
    print("5. Data completeness check:")
    missing = conn.execute("""
        WITH expected AS (
            SELECT DISTINCT image_id FROM image_token_distributions
        )
        SELECT
            image_id,
            COUNT(DISTINCT token_position) as positions
        FROM image_token_distributions
        GROUP BY image_id
        HAVING COUNT(DISTINCT token_position) != 256;
    """).fetchall()

    if missing:
        print(f"   ✗ FAILED: Some images missing positions:")
        for row in missing:
            print(f"     {row}")
        all_passed = False
    else:
        print("   ✓ All images have all 256 positions\n")

    conn.close()

    if all_passed:
        print("=" * 50)
        print("✅ ALL VALIDATION CHECKS PASSED")
        print("=" * 50)
        return True
    else:
        print("=" * 50)
        print("❌ SOME VALIDATION CHECKS FAILED")
        print("=" * 50)
        return False

def main():
    db_path = "vlm_analysis.db"
    passed = validate_database(db_path)
    sys.exit(0 if passed else 1)

if __name__ == "__main__":
    main()
```

### 3. Code Style Guidelines

**Match existing repository style**:
- Use type hints for all function signatures
- Follow existing naming conventions in `vlm_tools.py`
- Clean, functional style (avoid unnecessary classes)
- Docstrings with Args/Returns sections
- Use existing imports where possible (torch, transformers, etc.)

**Example of existing style** (from `vlm_tools.py`):
```python
def dissect_image(
    image_path: str,
    model: Gemma3ForConditionalGeneration,
    processor: AutoProcessor
) -> None:
    """
    Given an image, run the image through Gemma 3's vision tower and multi modal projector.
    Given those 256 image tokens, unembed them.
    Then, run frequency counts across the tokens that occur from the unembed.
    """
    image_tokens = extract_image_tokens(image_path, model, processor)
    vocabulary_tokens = unembed_to_vocabulary(image_tokens, model, processor)
    analyze_token_frequencies(vocabulary_tokens).show(max_rows=256)
```

### 4. Configuration

- **Top-p threshold**: 0.9 (configurable parameter, but default to 0.9)
- **Database path**: `vlm_analysis.db` (in repo root)
- **Images to process**: All 6 images in `images/` directory

### 5. Validation & Testing

After implementation, run `uv run python validate_database.py` which performs:

1. **Row count check**: Verify each image has data (256-50000 tokens total per image)
2. **Probability sum check**: Verify all positions have cumulative probability ≥0.9 and ≤1.0
3. **Ranking consistency check**: Verify probability_rank matches actual probability ordering
4. **Embedding vectors check**: Verify all 1,536 embeddings stored (6 images × 256 positions)
5. **Data completeness check**: Verify all images have all 256 positions

The validation script will exit with code 0 if all checks pass, code 1 if any fail.

### 6. Expected Output

After running `uv run python populate_database.py`, you should see something like:

```
Processing images/monocolor.png...
  Inserted 1,247 total tokens
Processing images/bicolor.png...
  Inserted 1,389 total tokens
Processing images/sunrise_mountain.jpg...
  Inserted 2,034 total tokens
Processing images/chicago_l.jpg...
  Inserted 1,876 total tokens
Processing images/code_snippet.png...
  Inserted 2,156 total tokens
Processing images/vision_language_model_data_flow.png...
  Inserted 1,998 total tokens

Database populated: vlm_analysis.db
```

And a `vlm_analysis.db` file ready for Phase 2 analysis.

## Deliverables

- [ ] `database.py` - Database operations (create schema, insert, query utilities)
- [ ] Extended `vlm_tools.py` - Add `extract_image_token_distributions` function
- [ ] `populate_database.py` - Executable script to populate database
- [ ] `validate_database.py` - Validation script that exits with code 0 on success, 1 on failure
- [ ] `vlm_analysis.db` - Populated database with all 6 images
- [ ] All validation checks pass

## Notes

- The database is queryable and will be used for Phase 2 positional analysis
- Storing raw embeddings enables future analyses without re-running inference
- The normalized schema allows flexible queries (e.g., "what tokens appear at position 0?", "what's the entropy at position 128?")
- This is a proof-of-concept with 6 images; schema supports scaling to larger datasets

## Success Criteria

1. ✅ Database created with correct schema
2. ✅ All 6 images processed and inserted
3. ✅ `validate_database.py` runs and exits with code 0 (all checks pass)
4. ✅ Rank 1 tokens match existing argmax implementation
5. ✅ Code style matches existing repository conventions
6. ✅ Database ready for Phase 2 analysis queries

## Pull Request Workflow

**IMPORTANT**: Only create a pull request if `validate_database.py` exits successfully (code 0).

1. Run `uv run python populate_database.py` to create and populate database
2. Run `uv run python validate_database.py` to verify data integrity
3. If validation fails:
   - Debug the issue
   - Fix the code
   - Re-run populate → validate
   - Iterate until validation passes
4. If validation passes:
   - Commit all changes with message: "feat: add top-p probability distribution database infrastructure"
   - Push to `feat/top-p-database` branch
   - Create pull request with:
     - Title: "feat: add top-p probability distribution database infrastructure"
     - Description: Include validation output + summary of implementation
   - Task complete ✅

**Autonomous execution**: Keep working through issues until PR is successfully created. Do not stop at first failure.
